src/Streamly/Coreutils/Uniq.hs

```diff
@@ -0,0 +1,152 @@
+                     , checkChar :: Int       -- take first checkChar number of characters for comparison (after above op)
+                     , ignoreCase :: Bool     -- whether to ignore case
+                     , unique :: Bool         -- print only unique lines
+                     , delimiter :: Char      -- to separate output lines
+                  }
+
+{-# INLINE defaultUniqOptions #-}
+
+defaultUniqOptions :: UniqOptions
+defaultUniqOptions = UniqOptions True False False 0 0 intMax False True '\n'
+
+
+-------------------------------------------------------------------------------
+-- helper functions for uniq
+-------------------------------------------------------------------------------
+
+
+-- Gives index=len is we have to skip the entire thing
+indexAfterSkippingNWords :: Int -> String -> Int
+indexAfterSkippingNWords n str = go 0 n
+  where
+    len = length str
+    go i 0 = i
+    go i _ | i >= (len - 1) = len
+    go i x = if isChar (str !! i) && isSpace (str !! (i + 1))
+             then go (i + 1) (x - 1)
+             else go (i + 1) x
+    isChar = not . isSpace
+
+
+-- slice offset length
+slice :: Int -> Int -> String -> String
+slice off len = take len . drop off
+
+
+compareIgnCase :: String -> String -> Bool
+compareIgnCase str1 str2 | length str1 /= length str2 = False
+compareIgnCase (x1:xs) (y1:ys) =
+    (toLower x1 == toLower y1) && compareIgnCase xs ys
+compareIgnCase [] [] = True
+
+
+compareWith :: (String -> String -> Bool) -> String -> String -> Bool
+compareWith eq s1 s2 = eq s1 s2
+
+
+-- Skip X terms
+-- Skip Y chars
+-- Take Z chars
+compareXYZ :: Bool -> Int -> Int -> Int -> String -> String -> Bool
+compareXYZ ignCase x y z str1 str2 =
+      compareWith compareFunc (slice (ix1 + y) z str1) (slice (ix2 + y) z str2)
+  where
+    ix1 = indexAfterSkippingNWords x str1
+    ix2 = indexAfterSkippingNWords x str2
+    compareFunc =
+        if ignCase
+        then compareIgnCase
+        else (==)
+
+
+splitOnNewLine :: (MonadIO m, IsStream t, Monad m) => t m Char -> t m String
+splitOnNewLine strm = S.splitOnSuffix (== '\n') FL.toList strm
+
+
+-- Skip X terms
+-- Skip Y chars
+-- Take Z chars
+getRepetition :: (IsStream t, Monad m) => Bool -> Int -> Int -> Int -> t m String -> t m (Int, String)
+getRepetition ignCase x y z = S.groupsBy (compareXYZ ignCase x y z)
+                              (FL.mkPureId (\(i, a) s -> if a == ""
+                                                         then (i + 1, s)
+                                                         else (i + 1, a)) (0, ""))
+
+
+onlyDuplicate :: (IsStream t, Monad m) => t m (Int, String) -> t m String
+onlyDuplicate = S.concatMap (\(i, x) -> if i > 1
+                                        then S.replicate i x
+                                        else S.nil)
+
+
+onlyRepeated :: (IsStream t, Monad m) => t m (Int, String) -> t m String
```

Comment:
Why is the input `t m (Int, String)`? It should be `t m String`?

Reply:
`uniqStrm`, `duplicateStrm`, `repeatedStrm` have signatures like `t m Char -> t m String`

If the input here would have been `t m String`, then I would also have to pass the `Bool -> Int -> Int -> Int` arguments for `getRepetition`.

Reply:
```
onlyRepeated :: (String -> String -> Bool) -> t m String -> t m (Int, String)
onlyRepeated comparator str = 
    S.concatMap (\(i, x) -> 
        if i > 1
        then S.replicate 1 x
        else S.nil)
    . getRpetition
```
?
Similarly, with `onlyDuplicate`.
This function is too specific to be a top-level function.
You can either change it like so or put it on the `where` clause of `uniqStrm`
Please do the same for `onlyUnique` and `onlyDuplicate` as well.

Reply:
okay. :+1: